MImperativeExpression not working in RAP E4

Looks like
`org.eclipse.e4.ui.workbench.renderers.swt.MenuManagerRenderer.updateVisibility(MenuManager, 
MMenuElement, ExpressionContext)` has not been upgraded when imperative expressions were introduced.

I tried to test it but
`mvn clean package` gave me
[ERROR] Internal error: org.eclipse.tycho.core.ee.UnknownEnvironmentException: Unknown OSGi execution environment: 'JavaSE-17'

No idea how to proceed.

--- a/bundles/org.eclipse.e4.ui.workbench.renderers.swt/src/org/eclipse/e4/ui/workbench/renderers/swt/MenuManagerRenderer.java
+++ b/bundles/org.eclipse.e4.ui.workbench.renderers.swt/src/org/eclipse/e4/ui/workbench/renderers/swt/MenuManagerRenderer.java
@@ -44,7 +44,6 @@ import org.eclipse.e4.ui.internal.workbench.ContributionsAnalyzer;
 import org.eclipse.e4.ui.internal.workbench.OpaqueElementUtil;
 import org.eclipse.e4.ui.internal.workbench.RenderedElementUtil;
 import org.eclipse.e4.ui.model.application.MApplication;
-import org.eclipse.e4.ui.model.application.ui.MCoreExpression;
 import org.eclipse.e4.ui.model.application.ui.MElementContainer;
 import org.eclipse.e4.ui.model.application.ui.MUIElement;
 import org.eclipse.e4.ui.model.application.ui.MUILabel;
@@ -1097,24 +1096,21 @@ MenuManagerEventHelper.getInstance()
 	 * @param element
 	 * @param evalContext
 	 */
-	public static void updateVisibility(MenuManager menuManager,
-			MMenuElement element, ExpressionContext evalContext) {
+	public static void updateVisibility(MenuManager menuManager, MMenuElement element, ExpressionContext evalContext) {
 		boolean current = element.isVisible();
 		boolean visible = true;
 		boolean evaluated = false;
 		if (element.getPersistedState().get(VISIBILITY_IDENTIFIER) != null) {
 			evaluated = true;
-			String identifier = element.getPersistedState().get(
-					VISIBILITY_IDENTIFIER);
+			String identifier = element.getPersistedState().get(VISIBILITY_IDENTIFIER);
 			Object rc = evalContext.eclipseContext.get(identifier);
 			if (rc instanceof Boolean) {
 				visible = ((Boolean) rc).booleanValue();
 			}
 		}
-		if (visible && (element.getVisibleWhen() instanceof MCoreExpression)) {
+		if (visible && element.getVisibleWhen() != null) {
 			evaluated = true;
-			visible = ContributionsAnalyzer.isVisible(
-					(MCoreExpression) element.getVisibleWhen(), evalContext);
+			visible = ContributionsAnalyzer.isVisible(element.getVisibleWhen(), evalContext);
 		}
 		if (evaluated && visible != current) {
 			element.setVisible(visible);

--- a/bundles/org.eclipse.e4.ui.workbench/src/org/eclipse/e4/ui/internal/workbench/ContributionsAnalyzer.java
+++ b/bundles/org.eclipse.e4.ui.workbench/src/org/eclipse/e4/ui/internal/workbench/ContributionsAnalyzer.java
@@ -12,6 +12,7 @@
 
 package org.eclipse.e4.ui.internal.workbench;
 
+import java.lang.annotation.Annotation;
 import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.HashSet;
@@ -22,12 +23,22 @@ import org.eclipse.core.expressions.Expression;
 import org.eclipse.core.expressions.ExpressionInfo;
 import org.eclipse.core.expressions.ReferenceExpression;
 import org.eclipse.e4.core.commands.ExpressionContext;
+import org.eclipse.e4.core.contexts.ContextInjectionFactory;
+import org.eclipse.e4.core.contexts.EclipseContextFactory;
 import org.eclipse.e4.core.contexts.IEclipseContext;
+import org.eclipse.e4.core.di.InjectionException;
+import org.eclipse.e4.core.di.InjectorFactory;
+import org.eclipse.e4.core.di.annotations.Evaluate;
+import org.eclipse.e4.core.di.suppliers.PrimaryObjectSupplier;
+import org.eclipse.e4.core.internal.contexts.ContextObjectSupplier;
+import org.eclipse.e4.core.internal.di.InjectorImpl;
+import org.eclipse.e4.core.services.contributions.IContributionFactory;
 import org.eclipse.e4.ui.model.application.MApplication;
 import org.eclipse.e4.ui.model.application.commands.MCommand;
 import org.eclipse.e4.ui.model.application.ui.MCoreExpression;
 import org.eclipse.e4.ui.model.application.ui.MElementContainer;
 import org.eclipse.e4.ui.model.application.ui.MExpression;
+import org.eclipse.e4.ui.model.application.ui.MImperativeExpression;
 import org.eclipse.e4.ui.model.application.ui.MUIElement;
 import org.eclipse.e4.ui.model.application.ui.basic.MTrimBar;
 import org.eclipse.e4.ui.model.application.ui.basic.MTrimElement;
@@ -46,6 +57,9 @@ import org.eclipse.emf.ecore.EObject;
 import org.eclipse.emf.ecore.util.EcoreUtil;
 
 public final class ContributionsAnalyzer {
+
+	private static final Object missingEvaluate = new Object();
+
 	public static void trace(String msg, Throwable error) {
 		Activator.trace("/trace/menus", msg, error); //$NON-NLS-1$
 	}
@@ -219,32 +233,44 @@ public final class ContributionsAnalyzer {
 		if (menuContribution.getVisibleWhen() == null) {
 			return true;
 		}
-		return isVisible((MCoreExpression) menuContribution.getVisibleWhen(), eContext);
+		return isVisible(menuContribution.getVisibleWhen(), eContext);
 	}
 
 	public static boolean isVisible(MToolBarContribution contribution, ExpressionContext eContext) {
 		if (contribution.getVisibleWhen() == null) {
 			return true;
 		}
-		return isVisible((MCoreExpression) contribution.getVisibleWhen(), eContext);
+		return isVisible(contribution.getVisibleWhen(), eContext);
 	}
 
 	public static boolean isVisible(MTrimContribution contribution, ExpressionContext eContext) {
 		if (contribution.getVisibleWhen() == null) {
 			return true;
 		}
-		return isVisible((MCoreExpression) contribution.getVisibleWhen(), eContext);
+		return isVisible(contribution.getVisibleWhen(), eContext);
 	}
 
-	public static boolean isVisible(MCoreExpression exp, final ExpressionContext eContext) {
-		final Expression ref;
-		if (exp.getCoreExpression() instanceof Expression) {
-			ref = (Expression) exp.getCoreExpression();
-		} else {
-			ref = new ReferenceExpression(exp.getCoreExpressionId());
-			exp.setCoreExpression(ref);
+	public static boolean isVisible(MExpression exp, final ExpressionContext eContext) {
+		if (exp instanceof MCoreExpression) {
+			MCoreExpression coreExpression = (MCoreExpression) exp;
+			return isCoreExpressionVisible(coreExpression, eContext);
+		} else if (exp instanceof MImperativeExpression) {
+			return isImperativeExpressionVisible((MImperativeExpression) exp, eContext);
 		}
-		// Creates dependency on a predefined value that can be "poked" by the evaluation
+
+		return true;
+	}
+
+	private static boolean isCoreExpressionVisible(MCoreExpression coreExpression, final ExpressionContext eContext) {
+		final Expression ref;
+		if (coreExpression.getCoreExpression() instanceof Expression) {
+			ref = (Expression) coreExpression.getCoreExpression();
+		} else {
+			ref = new ReferenceExpression(coreExpression.getCoreExpressionId());
+			coreExpression.setCoreExpression(ref);
+		}
+		// Creates dependency on a predefined value that can be "poked" by
+		// the evaluation
 		// service
 		ExpressionInfo info = ref.computeExpressionInfo();
 		String[] names = info.getAccessedPropertyNames();
@@ -255,9 +281,62 @@ public final class ContributionsAnalyzer {
 		try {
 			ret = ref.evaluate(eContext) != EvaluationResult.FALSE;
 		} catch (Exception e) {
-			trace("isVisible exception", e); //$NON-NLS-1$
+			if (DEBUG) {
+				trace("isVisible exception", e); //$NON-NLS-1$
+			}
 		}
 		return ret;
+	}
+
+	private static boolean isImperativeExpressionVisible(MImperativeExpression exp, final ExpressionContext eContext) {
+		Object imperativeExpressionObject = exp.getObject();
+		if (imperativeExpressionObject == null) {
+			IContributionFactory contributionFactory = eContext.eclipseContext.get(IContributionFactory.class);
+			Object newImperativeExpression = contributionFactory.create(exp.getContributionURI(),
+					eContext.eclipseContext);
+			exp.setObject(newImperativeExpression);
+			imperativeExpressionObject = newImperativeExpression;
+		}
+
+		if (imperativeExpressionObject == null) {
+			return false;
+		}
+
+		Object result = null;
+
+		IEclipseContext staticContext = EclipseContextFactory.create("Evaluation-Static");//$NON-NLS-1$
+		staticContext.set(MImperativeExpression.class, exp);
+		try {
+			if (exp.isTracking()) {
+				result = invoke(imperativeExpressionObject, Evaluate.class, eContext.eclipseContext, staticContext,
+						missingEvaluate);
+			} else {
+				result = ContextInjectionFactory.invoke(imperativeExpressionObject, Evaluate.class,
+						eContext.eclipseContext, staticContext, missingEvaluate);
+			}
+		} finally {
+			staticContext.dispose();
+		}
+
+		if (result == missingEvaluate) {
+			String className = "null";//$NON-NLS-1$
+			if (imperativeExpressionObject != null) {
+				className = imperativeExpressionObject.getClass().getName();
+			}
+			throw new IllegalStateException(
+					"There is no method annotated with @Evaluate in the " + className + " imperative expression class " //$NON-NLS-1$ //$NON-NLS-2$
+			);
+		}
+		return (boolean) result;
+	}
+
+	final private static InjectorImpl injector = (InjectorImpl) InjectorFactory.getDefault();
+
+	static private Object invoke(Object object, Class<? extends Annotation> qualifier, IEclipseContext context,
+			IEclipseContext localContext, Object defaultValue) throws InjectionException {
+		PrimaryObjectSupplier supplier = ContextObjectSupplier.getObjectSupplier(context, injector);
+		PrimaryObjectSupplier tempSupplier = ContextObjectSupplier.getObjectSupplier(localContext, injector);
+		return injector.invoke(object, qualifier, defaultValue, supplier, tempSupplier, false, true);
 	}
 
 	public static void addMenuContributions(final MMenu menuModel,

